internals-js/src/precompute.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  assert,
+  baseType,
+  CompositeType,
+  federationMetadata,
+  FieldDefinition,
+  collectTargetFields,
+  InterfaceType,
+  ObjectType,
+  Schema,
+  DirectiveDefinition
+} from ".";
+
+export const computeShareables = (schema: Schema) => {
+  const metadata = federationMetadata(schema);
+  return computeAllFieldsWithDirective(schema, metadata?.isFed2Schema() ? metadata.shareableDirective() : undefined, true);
+};
+
+export const computeKeys = (schema: Schema) => {
+  const metadata = federationMetadata(schema);
+  return computeAllFieldsWithDirective(schema, metadata?.isFed2Schema() ? metadata.keyDirective() : undefined, false);
+};
+
+function computeAllFieldsWithDirective(schema: Schema, directive: DirectiveDefinition<any> | undefined, includeOtherFieldsOnType: boolean): (field: FieldDefinition<CompositeType>) => boolean {
```

Comment:
In general, I think this method doesn't really do what the name implies and is still too tied to shareable (and there is still a bunch of mention to shareable which highlight this function is probably the wrong thing to use to compute keys).

Overall, the way I'd organise this to first have a method that precompute the list of field (coordinates) that appears in any `@key`. That code should not bother with provides or even have to care about extensions in any ways. It's probably something along the lines of just:
```typescript
metadata.keyDirective().applications().flapMap((key) => collectTargetFields(...)).map((f) => f.coordinate);
```
In fact, that method should probably be made generic to also allow pre-compute all the requires and provides fields (separately) since as mentioned in another comment, we should handle those consistently with how key fields are handled for override.

Then shareability can be defined using the precomputed list of keys since shareable fields are fields that either:
- key fields
- or provided fields
- or field with @shareable or part of a type with @shareable.

Reply:
I restored your original version of computeShareables. Let me know if I got the right message on computeKeys. If not, we should talk about it in person tomorrow.